services/serviceutils/src/main/java/tech/pegasys/teku/service/serviceutils/ScheduledExecutorAsyncRunner.java

```diff
@@ -0,0 +1,116 @@
+import tech.pegasys.teku.metrics.TekuMetricCategory;
+import tech.pegasys.teku.util.async.AsyncRunner;
+import tech.pegasys.teku.util.async.SafeFuture;
+
+class ScheduledExecutorAsyncRunner implements AsyncRunner {
+  private static final Logger LOG = LogManager.getLogger();
+  private final AtomicBoolean shutdown = new AtomicBoolean(false);
+  private final ScheduledExecutorService scheduler;
+  private final ExecutorService workerPool;
+
+  ScheduledExecutorAsyncRunner(
+      final ScheduledExecutorService scheduler, final ExecutorService workerPool) {
+    this.scheduler = scheduler;
+    this.workerPool = workerPool;
+  }
+
+  static AsyncRunner create(
+      final String name, final int maxThreads, final MetricsSystem metricsSystem) {
+    final ScheduledExecutorService scheduler =
+        Executors.newSingleThreadScheduledExecutor(
+            new ThreadFactoryBuilder()
+                .setNameFormat(name + "-async-scheduler-%d")
+                .setDaemon(true)
+                .build());
+    final ThreadPoolExecutor workerPool =
+        new ThreadPoolExecutor(
+            1,
+            maxThreads,
+            60,
+            TimeUnit.SECONDS,
+            new SynchronousQueue<>(),
+            new ThreadFactoryBuilder().setNameFormat(name + "-async-%d").setDaemon(true).build());
+
+    metricsSystem.createIntegerGauge(
+        TekuMetricCategory.EXECUTOR,
+        name + "_queue_size",
+        "Current size of the executor task queue",
+        () -> workerPool.getQueue().size());
+    metricsSystem.createIntegerGauge(
+        TekuMetricCategory.EXECUTOR,
+        name + "_thread_pool_size",
+        "Current number of threads in the executor thread pool",
+        workerPool::getPoolSize);
+    metricsSystem.createIntegerGauge(
+        TekuMetricCategory.EXECUTOR,
+        name + "_thread_active_count",
+        "Current number of threads executing tasks for this executor",
+        workerPool::getActiveCount);
+
+    return new ScheduledExecutorAsyncRunner(scheduler, workerPool);
+  }
+
+  @Override
+  public <U> SafeFuture<U> runAsync(final Supplier<SafeFuture<U>> action) {
+    return runTask(action, workerPool::execute);
+  }
+
+  @Override
+  @SuppressWarnings("FutureReturnValueIgnored")
+  public <U> SafeFuture<U> runAfterDelay(
+      final Supplier<SafeFuture<U>> action, final long delayAmount, final TimeUnit delayUnit) {
+    return runTask(
+        action, task -> scheduler.schedule(() -> workerPool.execute(task), delayAmount, delayUnit));
+  }
+
+  @Override
+  public void shutdown() {
+    // All threads are daemon threads so don't wait for them to actually stop
+    shutdown.set(true);
+    scheduler.shutdownNow();
+    workerPool.shutdownNow();
+  }
+
+  private <U> SafeFuture<U> runTask(
+      final Supplier<SafeFuture<U>> action, final Consumer<Runnable> scheduler) {
+    if (shutdown.get()) {
+      LOG.debug("Ignoring async task because shutdown is in progress");
+      return new SafeFuture<>();
```

Comment:
Should we not be returning a failed future here?

Reply:
No we're deliberately just not executing the task.  This is the same as if the task had been submitted before shutdown but hadn't yet been executed - we'd shutdown without ever completing the future.  Otherwise we have to go round the code base ignoring `RejectedExecutionException` everywhere.